types/validator_set.go

```diff
@@ -496,12 +495,37 @@ func (vals *ValidatorSet) VerifyFutureCommit(newSet *ValidatorSet, chainID strin
 	}
 
 	if oldVotingPower <= oldVals.TotalVotingPower()*2/3 {
-		return cmn.NewError("Invalid commit -- insufficient old voting power: got %v, needed %v",
-			oldVotingPower, oldVals.TotalVotingPower()*2/3+1)
+		return errTooMuchChange{oldVotingPower, oldVals.TotalVotingPower()*2/3 + 1}
```

Comment:
Once `errTooMuchChange` is returned, the verification algorithm would try to find out a middle point so that the validator set change is not too much for both prior one and the back one. 

But, if there is no point which can meet the above confidition, how to do with that situation?

Suppose now there are four validators(marked as A, B, C, D) and their voting power are 100. On height 1000, there is a delegator which delegate 10000 tokens to validator A. Then the voting power of validator A is 10100. In this situation, the verification algorithm will still report error.

What do you think about this situation?

Reply:
> the verification algorithm would try to find out a middle point so that the validator set change is not too much for both prior one and the back one.

If it never finds one where the valset change is small enough, it eventually gets back to a block where the validator set actually changed, and it can verify that change directly, thanks to the NextValidatorSetHash in the header.

The test in https://github.com/tendermint/tendermint/pull/3171/files#diff-0e9c3f682613cbf3fbae4b768e9c8fdeR78 tests a case where we have 5 validators, and in one block replace all of them at once with a new 5. Because the old validator set signs a header with the NextValidatorSetHash before the new validator set takes over, we can verify it successfully. 

If it weren't for NextValidatorSetHash, we would either have to cap the validator set changes in a block, or execute blocks before proposing them.

Reply:
Great! Make sense!